File: python/n202_HappyNumber.py
# https://leetcode-cn.com/problems/happy-number/solution/
class Solution:
    def isHappy(self, n: int) -> bool:
        def cal_num(num):
            res = 0
            while num > 0:
                res += (num % 10) ** 2
                num = num // 10
            return res
            
        slow = cal_num(n)
        fast = cal_num(n)
        fast = cal_num(fast)
        while slow != fast:
            slow = cal_num(slow)
            fast = cal_num(fast)
            fast = cal_num(fast)
        if slow == 1:
            return True
        else:
            return False
                   
            
if __name__ == "__main__":
    solu = Solution()
    print(solu.isHappy(897))

File: cpp/n202_HappyNumber.cpp
// https://leetcode-cn.com/problems/happy-number/solution/
#include <iostream>
#include <vector>
#include <algorithm>
#include <math.h>
using namespace std;

class Solution
{
public:
    bool HappyNumber(int num)
    {
        int slow = cal_num(num);
        int fast = cal_num(num);
        fast = cal_num(fast);
        while (slow != fast) 
        {
            slow = cal_num(slow);
            fast = cal_num(fast);
            fast = cal_num(fast);
        }
        if (slow == 1) return true;
        else return false;
    }
    
    int cal_num(int num)
    {
        int res = 0;
        while (num > 0) 
        {
            res += pow((num % 10), 2);
            num = num / 10;
        }
        return res;
    }
};

int main()
{
    Solution solu;
    cout << solu.HappyNumber(19) << endl;
}